pageserver/src/storage_sync.rs

```diff
@@ -129,1441 +114,661 @@
-
-    for (sync_id, local_timeline) in local_timeline_files {
-        let TimelineLocalFiles(local_metadata, local_files) = local_timeline;
-        match index.timeline_entry_mut(&sync_id) {
-            Some(remote_timeline) => {
-                let (timeline_status, awaits_download) = compare_local_and_remote_timeline(
-                    &mut new_sync_tasks,
-                    sync_id,
-                    local_metadata,
-                    local_files,
-                    remote_timeline,
-                );
-                match local_timeline_init_statuses
-                    .0
-                    .entry(sync_id.tenant_id)
-                    .or_default()
-                    .entry(sync_id.timeline_id)
-                {
-                    hash_map::Entry::Occupied(mut o) => {
-                        {
-                            // defensive check
-                            warn!(
-                                "Overwriting timeline init sync status. Status {timeline_status:?}, timeline {}",
-                                sync_id.timeline_id
-                            );
-                        }
-                        o.insert(timeline_status);
-                    }
-                    hash_map::Entry::Vacant(v) => {
-                        v.insert(timeline_status);
-                    }
+            match upload_result {
+                Ok(()) => {
+                    break;
                 }
+                Err(e) => {
+                    let retries = task.retries.fetch_add(1, Ordering::SeqCst);
 
-                remote_timeline.awaits_download = awaits_download;
-            }
-            None => {
-                // TODO (rodionov) does this mean that we've crashed during tenant creation?
-                //  is it safe to upload this checkpoint? could it be half broken?
-                warn!(
-                    "marking {} as locally complete, while it doesnt exist in remote index",
-                    sync_id
-                );
-                new_sync_tasks.push_back((
-                    sync_id,
-                    SyncTask::upload(LayersUpload {
-                        layers_to_upload: local_files,
-                        uploaded_layers: HashMap::new(),
-                        metadata: Some(local_metadata.clone()),
-                    }),
-                ));
-                local_timeline_init_statuses
-                    .0
-                    .entry(sync_id.tenant_id)
-                    .or_default()
-                    .insert(
-                        sync_id.timeline_id,
-                        LocalTimelineInitStatus::LocallyComplete(local_metadata),
+                    error!(
+                        "failed to perform remote task {}, will retry (attempt {}): {:?}",
+                        task.op, retries, e
                     );
+
+                    exponential_backoff(
```

Comment:
How do we handle task shutdown here?

Reply:
addressed by https://github.com/neondatabase/neon/pull/2921